rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,716 @@
+renew their certs on time.
+
+These two expirations create some confusion, and do not allow for an obvious
+method to deny a bot access, aside from creating a lock.
+
+We would like to solve two expiration use cases:
+1. We should be able to prevent all bot resource access after a certain date,
+   including renewals, in a way that allows the bot to be resumed later if
+   desired. (I.e. the token resource must still exist.)
+
+   Locks may accomplish this, but some centralized management in the token
+   resource would be convenient.
+
+2. We should be able to prevent bot rejoins after a certain date, to control
+   rejoining conditions in tandem with the rejoin counter. The
+   `spec.rejoining.must_rejoin_before` field may accomplish this.
+
+TODO: Ensure this is solved and not confusing.
+
+#### Keypair Rotation
+
+Given the long-lived nature of the keypair credential, it's important to support
+rotation without bot downtime. Ideally, it should be possible to initiate a
+rotation from either the server (e.g. by setting a `rotate_on_next_renewal` flag
+on the token/bot instance) or `tbot` client.
+
+TODO: Expand this.
+
+#### Remaining Downsides
+
+- Repairing a bot that has exhausted all of its rejoins is still a semi-manual
+  process. It is significantly easier, and does not necessarily require any
+  changes on the impacted bot node itself, but is still annoying. Users can opt
+  out of this by setting `.spec.rejoining.unlimited=true`, but this has obvious
+  security implications.
+
+- Effort required to configure IaC / Terraform is still fairly high, even if
+  reduced.
+
+### Other Supporting Improvements
+
+Alongside `bound-keypair` joining, we have several UX proposals to further
+improve the usability of non-delegated joining.
+
+#### Longer-Lived Bots
+
+The renewable identity's 24 hour maximum TTL is too restrictive and should be
+lengthened. We propose raising this limit to 7 days, but keeping the default
+(1hr) the same.
+
+#### Bot Instance Locking
+
+When we introduced bot instances in
+[RFD0162](0162-machine-id-token-join-method-bot-instance.md), it allowed many
+bot instances to join under a single bot (and Teleport user). Generation counter
+checks were moved out of a bot's user and into bot instances, however when a
+generation counter mismatch occurs, the resulting lock is still filed against
+the user as a whole. This means a generation counter lockout in one instance can
+easily impact all other instances under the same bot.
+
+We can introduce several new lock targets to address this:
+
+- Bot instance UUID locking: prevent access by a particular instance. This will
+  not lock bots that have since reauthenticated and received a new bot instance
+  UUID.
+
+- Join token locking: locks all bot instances that joined with a particular
+  token. This may require introduction of a new certificate field to track the
+  exact join token used.
+
+- Public key locking: locks bots joining with a particular public key. A
```

Comment:
Another option would be to allow a bot to be limited to only a single instance. If a second instance joins, the first one is revoked (by design).

Another option would be to only allow rejoining if all instances are rejoining at once. Say after a power outage on a site with multiple instances. The logic would validate that there are other instances in the same bot, and only allow a generation increase if they all 'agree'.

Reply:
I've tried to further clarify the approach here in the "bot lifecycle" section since it was vague and wasn't fully written down, and actually pulling on this thread led to introducing "Join State Documents" in the new "Preventing Credential Duplication" section.

The short version is, we'll only allow one active bot instance per `(bot, bound-keypair token)` pair. If a rejoin occurs, the previous instance will not be allowed to refresh its certificates any further and will need to rejoin - and might additionally get locked out. If a bot's keypair is cloned and 2 clients start competing to rejoin, we'll detect this and lock all bots using the join token, similar to generation counter lockouts today.